File: rental_platform/rental_platform/report/customer_approvals/customer_approvals.py
import frappe
from frappe import _

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "fieldname": "customer",
            "label": _("Customer ID"),
            "fieldtype": "Link",
            "options": "Customer",
            "width": 120
        },
        {
            "fieldname": "customer_name",
            "label": _("Customer Name"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "email",
            "label": _("Email"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "mobile_no",
            "label": _("Mobile Number"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "creation",
            "label": _("Registration Date"),
            "fieldtype": "Datetime",
            "width": 150
        },
        {
            "fieldname": "portal_approval_status",
            "label": _("Approval Status"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "actions",
            "label": _("Actions"),
            "fieldtype": "Data",
            "width": 200
        }
    ]

def get_data(filters):
    conditions = []
    if filters and filters.get("status"):
        conditions.append(f"portal_approval_status = '{filters.get('status')}'")
        
    where_clause = " AND ".join(conditions) if conditions else "1=1"
        
    customers = frappe.db.sql(f"""
        SELECT
            name as customer,
            customer_name,
            custom_email as email,
            mobile_no,
            creation,
            portal_approval_status
        FROM `tabCustomer`
        WHERE {where_clause}
        ORDER BY creation DESC
    """, as_dict=1)
    
    for c in customers:
        if not c.portal_approval_status:
            c.portal_approval_status = "Approved"
        # We inject buttons in JS using a formatter, so we just return the row
        c.actions = ""
        
    return customers

File: rental_platform/rental_platform/report/customer_approvals/customer_approvals.js
frappe.query_reports["Customer Approvals"] = {
	"filters": [
		{
			"fieldname":"status",
			"label": __("Approval Status"),
			"fieldtype": "Select",
			"options": "\nPending\nApproved\nRejected",
			"default": "Pending"
		}
	],
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (column.fieldname == "actions") {
		    if (data.portal_approval_status === "Pending") {
			    return `
			        <button class="btn btn-xs btn-success" onclick="frappe.custom_approve_customer('${data.customer}')">Approve</button>
			        <button class="btn btn-xs btn-danger" onclick="frappe.custom_reject_customer('${data.customer}')">Reject</button>
			    `;
			} else {
			    return `<span>${data.portal_approval_status}</span>`;
			}
		}
		
		if (column.fieldname == "portal_approval_status") {
		    if (data.portal_approval_status === "Pending") {
		        return `<span class="indicator-pill orange">${data.portal_approval_status}</span>`;
		    } else if (data.portal_approval_status === "Approved") {
		        return `<span class="indicator-pill green">${data.portal_approval_status}</span>`;
		    } else {
		        return `<span class="indicator-pill red">${data.portal_approval_status}</span>`;
		    }
		}
		return value;
	}
};

frappe.custom_approve_customer = function(customer_id) {
    frappe.confirm('Are you sure you want to approve this customer?', () => {
        frappe.call({
            method: 'rental_platform.web_api.customer_portal_auth.update_approval_status',
            args: {
                customer_id: customer_id,
                status: 'Approved'
            },
            callback: function(r) {
                if (r.message && r.message.success) {
                    frappe.show_alert({message:__('Customer Approved'), indicator:'green'});
                    frappe.query_report.refresh();
                } else {
                    frappe.msgprint(r.message ? r.message.message : 'Error approving customer');
                }
            }
        });
    });
};

frappe.custom_reject_customer = function(customer_id) {
    frappe.confirm('Are you sure you want to reject this customer?', () => {
        frappe.call({
            method: 'rental_platform.web_api.customer_portal_auth.update_approval_status',
            args: {
                customer_id: customer_id,
                status: 'Rejected'
            },
            callback: function(r) {
                if (r.message && r.message.success) {
                    frappe.show_alert({message:__('Customer Rejected'), indicator:'red'});
                    frappe.query_report.refresh();
                } else {
                    frappe.msgprint(r.message ? r.message.message : 'Error rejecting customer');
                }
            }
        });
    });
};
